src/neuroglancer/sliceview/chunked_graph/backend.ts

```diff
@@ -0,0 +1,334 @@
+/**
+ * @license
+ * Copyright 2018 The Neuroglancer Authors
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import debounce from 'lodash/debounce';
+import {cancelChunkDownload, startChunkDownload, withChunkManager, ChunkRenderLayerBackend} from 'neuroglancer/chunk_manager/backend';
+import {ChunkPriorityTier, ChunkState} from 'neuroglancer/chunk_manager/base';
+import {SharedDisjointUint64Sets} from 'neuroglancer/shared_disjoint_sets';
+import {SliceViewChunk, SliceViewChunkSourceBackend} from 'neuroglancer/sliceview/backend';
+import {SliceViewRenderLayer as SliceViewRenderLayerInterface, filterVisibleSources, SliceViewBase, TransformedSource} from 'neuroglancer/sliceview/base';
+import {CHUNKED_GRAPH_LAYER_RPC_ID, CHUNKED_GRAPH_SOURCE_UPDATE_ROOT_SEGMENTS_RPC_ID, ChunkedGraphChunkSource as ChunkedGraphChunkSourceInterface, ChunkedGraphChunkSpecification, RENDER_RATIO_LIMIT} from 'neuroglancer/sliceview/chunked_graph/base';
+import {Uint64Set} from 'neuroglancer/uint64_set';
+import {vec3, vec3Key} from 'neuroglancer/util/geom';
+import {Uint64} from 'neuroglancer/util/uint64';
+import {registerRPC, registerSharedObject, RPC, SharedObjectCounterpart} from 'neuroglancer/worker_rpc';
+import {WatchableValueInterface} from 'neuroglancer/trackable_value';
+
+import * as vector from 'neuroglancer/util/vector';
+import {isBaseSegmentId} from 'neuroglancer/datasource/graphene/base';
+
+export class ChunkedGraphChunk extends SliceViewChunk {
+  backendOnly = true;
+  source: ChunkedGraphChunkSource|null = null;
+  mappings: Map<string, Uint64[]|null>|null = null;

```

Comment:
In Neuroglancer the existing framework for downloading and caching chunks is based on the assumption that they are immutable and cacheable and basically correspond to a single (or small fixed number) of network requests.

ChunkedGraphChunk would fit better with the existing framework if for each (level, grid_pos, root_id) there were a separate chunk, rather than the current system where there is a single chunk for each (level, grid_pos).  Then the ChunkedGraphSource would not depend on the rootSegments at all, and the ChunkedGraphLayer would be responsible for requesting chunks for the current set of root segments.

I imagine one reason for the current design of ChunkedGraphChunk is to be able to re-use the existing mechanism in SliceView for deciding which chunks to load, and at the time graphene was first implemented in the seung branch there wasn't a viable alternative.

However, note that the logic in SliceView has been factored out so that it can be used mostly independently --- see `forEachPlaneIntersectingVolumetricChunk` defined in sliceview/base.ts.

Reply:
It seems we had some intentions to make bulk /leaves requests for multiple segments within a region but we never used it in neuroglancer and we don't see a need for it currently so your proposition sounds good.

Could you guide me on the change needed to make ChunkedGraphLayer request chunks per root segment?

I see GrapheneMultiscaleVolumeChunkSource.getChunkedGraphSources which creates a single source which references visibleSegments. That is passed to the frontend ChunkedGraphLayer which has getSources method which we just return this.sources for.  Then on the backend ChunkedGraphLayer, we have filterVisibleSources which has some hacks I was hoping could be replaced.

So one guess is that we no longer generate the sources immediately on layer creation and getSources creates a new list of sources based on the current visible segments?

Reply:
There should still be one ChunkSource per chunk graph level, but its chunks should be keyed by position+root id, rather than just position (SliceViewChunkSource keys chunks by position only, but since this will be operating separately from SliceView you don't have to do that here).  What you describe of having a separate chunk source per root id could work, but it is simpler to just use a single chunk source.

Reply:
ChunkedGraphChunkSource.getChunk takes a key chunkGridPosition, it generates a key `vec3Key(chunkGridPosition)` creates a chunk passing it all the root segments and returns it. Do I need to change my getChunk to `getChunk(chunkGridPosition: Float32Array, segment: Uint64) {`?

Reply:
Yeah that's what I have in mind.

Reply:
Would ChunkedGraphLayer still extend SliceViewRenderLayer or do I need to replace SliceViewRenderLayer and all other SliceView classes/factor it out.

Reply:
you also mentioned above the "dummy renderlayer" that could be attached to a slice view which seems applicable to the chunked graph layer.

Reply:
@jbms I just added a commit that makes I think the changes to ChunkedGraphChunk that you were expecting, though I wasn't sure how to trigger what I called `handleChunkedGraphLayer`. I did a crude hack to get something working. I'm not sure what changes I should be making on the front end. I'm not sure what I should be doing as far as a ChunkManager as well. 

One thing I noticed is that `ChunkedGraphChunkSource.computeChunkBounds` is essentially identical to `VolumeChunkSource`.

I think the changes once you get inside ChunkedGraphChunk are a lot cleaner than our previous implementation.

Reply:
Yes --- I think `computeChunkBounds` could be factored out and made a non-member function --- then it wouldn't have to be duplicated.


See my comment on your commit here:

https://github.com/google/neuroglancer/commit/35916543e1e8877dea4af598fa3cba892125fd1f#r71754865

I think with a relatively small change the integration of ChunkedGraphLayer can be cleaner.

Overall this looks very close to being ready.